Sources/CacheAPI-postgres.php

```diff
@@ -41,30 +42,67 @@ public function __construct()
 	 */
 	public function connect()
 	{
-		global $db_prefix, $db_connection;
-
-		pg_prepare($db_connection, '', 'SELECT 1
+		$result = pg_query_params($this->db_connection, 'SELECT 1
 			FROM   pg_tables
 			WHERE  schemaname = $1
-			AND    tablename = $2');
-
-		$result = pg_execute($db_connection, '', array('public', $db_prefix . 'cache'));
+			AND    tablename = $2',
+			array(
+				'public',
+				$this->db_prefix . 'cache',
+			)
+		);
 
 		if (pg_affected_rows($result) === 0)
-			pg_query($db_connection, 'CREATE UNLOGGED TABLE ' . $db_prefix . 'cache (key text, value text, ttl bigint, PRIMARY KEY (key))');
+		{
+			pg_query($this->db_connection, 'CREATE UNLOGGED TABLE ' . $this->db_prefix . 'cache (key text, value text, ttl bigint, PRIMARY KEY (key))');
+
+			$this->prepareQuery(
+				'smf_cache_get_data',
+				'SELECT value FROM ' . $this->db_prefix . 'cache WHERE key = $1 AND ttl >= $2 LIMIT 1'
+			);
+			$this->prepareQuery(
+				'smf_cache_put_data',
+				'INSERT INTO ' . $this->db_prefix . 'cache(key,value,ttl) VALUES($1,$2,$3)
+				ON CONFLICT(key) DO UPDATE SET value = $2, ttl = $3'
+			);
+			$this->prepareQuery(
+				'smf_cache_delete_data',
+				'DELETE FROM ' . $this->db_prefix . 'cache WHERE key = $1'
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Stores a prepared SQL statement, ensuring that it's not done twice.
+	 *
+	 * @param string $stmtname
+	 * @param string $query
+	 */
+	private function prepareQuery($stmtname, $query)
+	{
+		$result = pg_query_params(
+			$this->db_connection,
+			'SELECT name FROM pg_prepared_statements WHERE name = $1',
+			array($stmtname)
+		);
```

Comment:
Show me a better way to check if the resource is allocated.

Reply:
There is no check needed, 
When the prepare run well,
Than came up with no error, 
When already exists he can up with an error which needs to be ignored.

Reply:
If I understand you correctly, @albertlast, you mean that instead of running this:
```
$this->prepareQuery(
 	'smf_cache_delete_data',
 	'DELETE FROM ' . $this->db_prefix . 'cache WHERE key = $1'
 );
```
... we should instead just run this:
```
@pg_prepare(
 	$this->db_connection,
 	'smf_cache_delete_data',
 	'DELETE FROM ' . $this->db_prefix . 'cache WHERE key = $1'
 );
```

Do I understand you correctly?

Reply:
No realy,
to centralize the prepare logic is fine,
but the check in form of a select is waste of time.

Reply:
The version of https://github.com/SimpleMachines/SMF2.1/pull/5969/commits/19c2abb7675ef60c322546f06ab27f7c0df53672
to combine all selects in one and
prepare only this how are not exists is a ok solution,
but the check is still usless so long you don't had a persitent connection only in this case a prepare query can exists or did i miss something?

Reply:
So the prepared statements only last as long as the connection? Would it be safe to forego the query for regular non-persistent connections?

Reply:
I short yes,
https://www.postgresql.org/docs/current/sql-prepare.html :
```
Prepared statements only last for the duration of the current database session.
When the session ends, the prepared statement is forgotten,
so it must be recreated before being used again.
```
reason why in the existing code a persistent check/special logic exists.